src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java

```diff
@@ -108,4 +149,84 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     protected String parserName() {
         return IdsQueryParser.NAME;
     }
+
+    public Query toQuery(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        ArrayList<BytesRef> ids = new ArrayList<>();
+
+        for (String value : this.ids) {
+            BytesRef ref = new BytesRef(value);
+            ids.add(ref);
+        }
+
+        if (ids.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+
+        if (types == null || types.isEmpty()) {
+            types = parseContext.queryTypes();
+        } else if (types.size() == 1 && Iterables.getFirst(types, null).equals("_all")) {
+            types = parseContext.mapperService().types();
+        }
+
+        TermsQuery query = new TermsQuery(UidFieldMapper.NAME, Uid.createTypeUids(types, ids));
+        query.setBoost(boost);
+        if (queryName != null) {
+            parseContext.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        // all fields can be empty or null
+        return null;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        int typeSize = in.readInt();
+        for (int i = 0; i < typeSize; i++) {
+            types.add(in.readString());
+        }
+        int valueSize = in.readInt();
+        for (int i = 0; i < valueSize; i++) {
+            ids.add(in.readString());
+        }
```

Comment:
we don't have a utility method to read and write a Collection right? that would save us few lines of code. Shall we add it instead of repeating loops?